wizard-js/source-parser.js

```diff
@@ -0,0 +1,167 @@
+'use strict';
+
+const deepEqual = require('deep-equal');
+const uniq = require('uniq');
+const { flatten } = require('array-flatten');
+
+const parsers = {
+    expressRoutes: require('./source-parsers/core/express-routes.js'),
+    cloudFunction: require('./source-parsers/core/cloud-functions.js'),
+    directInvocation: require('./source-parsers/core/direct-invocation.js'),
+    wrappedMain: require('./source-parsers/extra/wrapped-main.js'),
+    wrappedDirectInvocation: require('./source-parsers/extra/wrapped-direct-invocation.js')
+};
+
+const getMethodChildren = (method) => {
+    if (!(method instanceof Object)) {
+        return [];
+    }
+
+    let children = [];
+    if (method.callee && method.callee.name) {
+        children.push(method.callee.name);
+    }
+
+    const props = Object.keys(method);
+    props.forEach((key) => {
+        if (key === 'loc' || !method[key]) {
+            return;
+        }
+
+        const value = method[key];
+        if (Array.isArray(value)) {
+            const newChildren = value.map(x => getMethodChildren(x))
+            children = children.concat(flatten(newChildren));
+        } else {
+            children = children.concat(getMethodChildren(value));
+        }
+    })
+
+    return children;
+}
+
+exports.getTopLevelMethods = (sourceTokens, sourcePath) => {
+    const expressMethods = parsers.expressRoutes.parse(sourceTokens, sourcePath);
+    const gcfMethods = parsers.cloudFunction.parse(sourceTokens, sourcePath);
+    const snippetMethods = parsers.directInvocation.parse(sourceTokens, sourcePath);
+    const wrappedMainMethods = parsers.wrappedMain.parse(sourceTokens, sourcePath);
+    const wrappedDirectInvocationMethods = parsers.wrappedDirectInvocation.parse(sourceTokens, sourcePath);
+    let allMethods = [
+        ...snippetMethods, 
+        ...expressMethods, 
+        ...gcfMethods, 
+        ...wrappedMainMethods,
+        ...wrappedDirectInvocationMethods];
+
+    // Remove duplicate methods
+    allMethods = uniq(allMethods, (x, y) => deepEqual(x, y) ? 0 : 1);
```

Comment:
Can we use `new Set(alllMethods)` instead of `uniq`?

Reply:
Unlike the current code, `new Set()` [doesn't seem to support](https://stackoverflow.com/questions/29759480/how-to-customize-object-equality-for-javascript-set/29783624) deep-equal comparison (which is necessary here).